File: social-service/src/main/java/com/eirsteir/coffeewithme/social/domain/friendship/Friendship.java
package com.eirsteir.coffeewithme.social.domain.friendship;

import com.eirsteir.coffeewithme.commons.domain.friendship.FriendRequestAcceptedEvent;
import com.eirsteir.coffeewithme.commons.domain.friendship.FriendRequestEvent;
import com.eirsteir.coffeewithme.commons.domain.user.UserDetails;
import com.eirsteir.coffeewithme.social.domain.user.User;
import io.eventuate.tram.events.publisher.ResultWithEvents;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.util.Collections;
import java.util.Date;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@Entity
@Table(name = "friendships")
@AssociationOverrides({
        @AssociationOverride(name = "id.requester", joinColumns = @JoinColumn(name ="requester_id")),
        @AssociationOverride(name = "id.addressee", joinColumns = @JoinColumn(name ="addressee_id"))
})
public class Friendship {

    @EmbeddedId
    private FriendshipId id;

    @Transient
    public User getRequester() {
        return id.getRequester();
    }

    @Transient
    public User getAddressee() {
        return id.getAddressee();
    }

    private FriendshipStatus status;

    @CreationTimestamp
    private Date createdDateTime;

    @UpdateTimestamp
    private Date updatedDateTime;

    public static ResultWithEvents<Friendship> createFriendRequest(Friendship friendship, UserDetails user) {

        FriendRequestEvent friendRequestEvent = new FriendRequestEvent(friendship.getAddressee().getId(), user);
        return new ResultWithEvents<>(friendship, Collections.singletonList(friendRequestEvent));
    }

    public static ResultWithEvents<Friendship> createFriendRequestAccepted(Friendship friendship, UserDetails user) {

        FriendRequestAcceptedEvent friendRequestAcceptedEvent = new FriendRequestAcceptedEvent(
                friendship.getRequester().getId(), user);
        return new ResultWithEvents<>(friendship, Collections.singletonList(friendRequestAcceptedEvent));
    }

    public static class FriendshipBuilder {
        private FriendshipId id = new FriendshipId();

        public FriendshipBuilder requester(User requester) {
            this.id.setRequester(requester);
            return this;
        }

        public FriendshipBuilder addressee(User addressee) {
            this.id.setAddressee(addressee);
            return this;
        }
    }
}

File: social-service/src/main/kotlin/com/eirsteir/coffeewithme/social/domain/friendship/Friendship.kt
package com.eirsteir.coffeewithme.social.domain.friendship

import com.eirsteir.coffeewithme.commons.domain.friendship.FriendRequestAcceptedEvent
import com.eirsteir.coffeewithme.commons.domain.friendship.FriendRequestEvent
import com.eirsteir.coffeewithme.commons.domain.user.UserDetails
import com.eirsteir.coffeewithme.social.domain.user.User
import io.eventuate.tram.events.publisher.ResultWithEvents
import lombok.Builder
import lombok.Data
import lombok.experimental.Accessors
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.util.*
import javax.persistence.*

@Entity
@AssociationOverrides(
    AssociationOverride(name = "id.requester", joinColumns = [JoinColumn(name = "requester_id")]),
    AssociationOverride(name = "id.addressee", joinColumns = [JoinColumn(name = "addressee_id")])
)
data class Friendship(
    @EmbeddedId
    val id: FriendshipId? = null,
    var status: FriendshipStatus? = null,
    @CreationTimestamp
    val createdDateTime: Date? = null,
    @UpdateTimestamp
    val updatedDateTime: Date? = null
) {
    @Transient
    val requester: User? = id?.requester

    @Transient
    val addressee: User? = id?.addressee

    constructor(requester: User, addressee: User, status: FriendshipStatus):
            this(id = FriendshipId(requester, addressee), status = status)

    companion object {
        fun createFriendRequest(
            friendship: Friendship, user: UserDetails
        ): ResultWithEvents<Friendship> {
            val friendRequestEvent = FriendRequestEvent(friendship.addressee?.id!!, user)
            return ResultWithEvents<Friendship>(friendship, listOf(friendRequestEvent))
        }

        fun createFriendRequestAccepted(
            friendship: Friendship, user: UserDetails
        ): ResultWithEvents<Friendship> {
            val friendRequestAcceptedEvent = FriendRequestAcceptedEvent(friendship.requester?.id!!, user)
            return ResultWithEvents<Friendship>(
                friendship, listOf(friendRequestAcceptedEvent)
            )
        }
    }
}